### DISCUSSION TITLE
Profile feature to automate the filling of a form

### DISCUSSION THREAD

**Discussion by Flocon29** [netbox-community/netbox#7506] [Ideas]
Currently we are working on adding a profile field in several models.
This profile field will be able to dynamically modify several other fields on a form, like the interface edition page below:

### Interface Edition

![interface_edit_profile](https://user-images.githubusercontent.com/18680214/149481701-4b13f680-c302-4ab7-b53a-1a815b47711f.gif)

Here, when a profile is selected, it dynamically updates fields like `VLANs` and also custom fields like `Link Mode`.

Custom-fields paired with config-contexts seemed to us to be a good solution to configure and customize our profiles on different models.
I will explain below how we set this up.

### Custom-Field Edition

First, our `Profile` field is defined as a classic `custom field`, but we needed to be able to differentiate it from others.
That's why we've added a `Profile Mode` checkbox that defines whether our `custom field` should behave like a profile on the edit page.

![image](https://user-images.githubusercontent.com/18680214/149123563-bdeac399-c41e-44bd-aeb2-35a477e9ae0a.png)

If the checkbox is marked, the above field named `Config context` replaces the `Choices` field. 
This new field defines the profiles available in the selection `dropdown-menu`.

### Config-Context Edition

And these profiles are `Config-Contexts` associated to a model (here `DCIM> interface`).
Then defined in JSON format, the various default values ​​to be assigned to this profile.

![image](https://user-images.githubusercontent.com/18680214/149482098-5620e564-ab5f-41c5-848a-85654210c288.png)

### Javascript

Then, when a profile is selected on the interface edition page, a javascript file will fill the form, according to the `context-data` defined for this profile.

An `Update Profile` checkbox has also been added to the `config-context` edit page.
If this box is marked during a Congfig-Context edition, a javascript file will update via the NetBox API all the elements for which this profile has already been defined.

Would you be interested in integrating this feature or an equivalent in Netbox?
If this is the case, we are open to share with you our developments in order to integrate and share these features with the community.

**Comment by julianze** [comment]
That would be great!
I see several use cases for this.

**Comment by Flocon29** [comment]
Hi @jeremystretch 

Would you be interested in this feature that we have developed using Custom-fields and Config-Context?
If so, we can help you integrate it natively into the Netbox application by sharing our developments with you.

**Reply by jeremystretch** [reply]
I think I see what you're after as far as the functionality, but IMO custom fields and context data are not the right tools for this. What you really need for this to be robust and maintainable is a separate model that can track the application of data to profiles without duplication.

I think the first step would be to open a feature request detailing specifically what functionality you'd like to see implemented, and its use case.

**Reply by Flocon29** [reply]
Hi, thanks for your answer.

Exactly, I totally agree with you.
We initially chose to use custom fields and context data in order to avoid developing a separate new model and so limit conflicts between our developments and those coming from the netbox community as much as possible, but we know this is not the most sustainable solution in terms of maintainability.
We simply had to slightly modify a few forms, then all our processing concerning the profiles is carried out from javascript files.

That's why today we are wondering this functionality would be interesting to integrate natively to Netbox
Because if the community can be interested by it, we would like to participate in its development and also integrate it directly into NetBox with better robustness and maintainability as you said.

So, we will open a request detailing the feature and its use case.

**Reply by geor-g** [reply]
@Flocon29 Thanks for doing, once done, please drop a link here. Thank you!

**Reply by Flocon29** [reply]
The feature request has been created here : https://github.com/netbox-community/netbox/issues/8371
